module-system/sov-state/src/codec/mod.rs

```diff
@@ -42,3 +43,38 @@ pub trait StateValueCodec<V> {
             .unwrap()
     }
 }
+
+pub trait EncodeLike<Ref: ?Sized, Target> {
+    fn encode_like(&self, borrowed: &Ref) -> Vec<u8>;
```

Comment:
I'm wondering if `EncodeLike` should be a marker trait instead. If both `Ref` and `Target` serialize the same way, why is a third method `encode_like` needed? Making this is a marker trait would then require `EncodeLike<..., ...> + StateValueCodec<...>` (or whatever codec trait we must use) in all places where we'll need to actually encode types. So it's more verbose.

This is non-blocking, feel free not to address this suggestion if you prefer the current design.

Reply:
Good question!

The motivation for this is that a type might have multiple encodings. For example `[u8;N]` should not ordinarily be encoded with a `length`, but it *can* be encoded that way. In this design, a StateCodec<T> specifies the canonical encoding for a type, and `EncodeLike<T, U>` specifies that T can be encoded using a (potentially different) format which is guaranteed to match the canonical format of `U`.

Reply:
Ah, makes sense! Big brain stuff.